### ISSUE TITLE
Does this library actually end up working reliable with Active Directory?

### ISSUE THREAD

**Issue by AzorianMatt** [cannatag/ldap3#1112]
So I realize that Microsoft is the king of taking an RFC and stomping all over it "because we run the show." For that reason, I'm wondering if anyone can confirm that this library in general entirely works with an Win Server 2012 AD LDAP? I know that the setup of this server was not likely performed by true Win Server admins so there is a whole plethora of things that don't seem right.

For example, when I query for all person objects, we aren't getting back the entire set of data and haven't identified a pattern as to why. The service account being used has the maximum permission set that can be given (allegedly). I can't get past an issue with creating new user objects because of this:

`<{'result': 19, 'description': 'constraintViolation', 'dn': '', 'message': '000021C8: AtrErr: DSID-03200E07, #1:\n\t0: 000021C8: DSID-03200E07, problem 1005 (CONSTRAINT_ATT_TYPE), data 0, Att 90290 (userPrincipalName)\n\x00', 'referrals': None, 'type': 'addResponse'}>`

Does anyone here have any experience with using this library against AD and whether or not it actually works? I don't know AD enough myself and our internal resources apparently don't either so I'm at a real road block here.

**Comment by Zamanry** [comment]
Yes, I’ve used it in my company’s tool https://github.com/CroweCybersecurity/ad-ldap-enum. Additionally, Fortra’s entire Impacket suite uses it as well which is one of the largest Active Directory/Windows assessing via Linux tools https://github.com/fortra/impacket

**Comment by zorn96** [comment]
@AzorianMatt it looks like you're just adding a user with an invalid principal name. without seeing the actual payload you're sending and also knowing domain policies, it's not obvious way.
potential issues are:
- invalid characters in the principal name
- trying to use a name that's already in use

**Comment by AzorianMatt** [comment]
@zorn96 thank you for your reply. I'll have to dig into the domain policies more to personally verify. Do you have any particularly helpful suggestions for what to look for in a domain policy?

As for the value I'm sending, I'm sending a standard email address with a period and @ sign being the only non-alpha characters. Is there a special encoding I must use for this that I have missed? I use values identical in format to many existing objects. I've even tried dropping the @domain indicator altogether but that just results in a will not perform exception.

**Comment by Zamanry** [comment]
> @zorn96 thank you for your reply. I'll have to dig into the domain policies more to personally verify. Do you have any particularly helpful suggestions for what to look for in a domain policy?
> 
> As for the value I'm sending, I'm sending a standard email address with a period and @ sign being the only non-alpha characters. Is there a special encoding I must use for this that I have missed? I use values identical in format to many existing objects. I've even tried dropping the @Domain indicator altogether but that just results in a will not perform exception.

If you look at my ad-ldap-enum raw code (crappy as it may be), you will see the ldap3 connection format I use to get Windows domain + user to work. I think that’s what you need

**Comment by AzorianMatt** [comment]
@Zamanry thanks. I did manage to finally get some movement and progress on the issue today. I have some working test cases now. When I shut down for the day though, I was back in a situation of getting non-failure hanging scenarios where the request just never completes but never kicks back any exception.

Apparently the server is being rebuilt tonight to something newer than 2012, so fingers crossed it doesn't come out more broken than before.
